### ISSUE TITLE
The int variable will never appear in the str dictionary

### ISSUE THREAD

**Issue by egmkang** [microsoft/pyright#2752]
```python
str_2_int = {}   # type: dict[str, int]
item = 0         # type: int

if item not in str_2_int:
    pass
```

in this code, dict's key has type `str`, but the `item` has type `int`. 

can pyright warn this case?

**Comment by erictraut** [comment]
This is not a type error, so no error is emitted. 

Are you suggesting that the expression `item not in str_2_int` should be considered an error because it is provably false? There are lots of provably-false or provably-true expressions used in `if` statements that are not necessarily indicative of a bug. A type checker needs to be conservative to avoid false positive type errors.

In this particular case, a type checker would need to make a number of tenuous assumptions to determine that this expression is always false. Unless this were a very common source of bugs, I don't think we could justify the risk of emitting false-positive errors.

I'm going to mark this feature request as closed for now. If it receives more upvotes (thumbs up) in the future, we might revisit this decision.

**Comment by egmkang** [comment]
@erictraut 

```
# python
if item not in str_2_int:
    pass

# C#
if (!str_2_int.ContainsKey(item)) {}
```

in C# or other languages, check key exists with another type usually get a compile error. in this case, i think expression `in` is equal to `ContainsKey`. if the compiler could give a warning (not an error), that will be very helpful.

**Comment by egmkang** [comment]
@erictraut 
```python

# case 1
if item not in str_2_int:
    pass

# case 2
if  str_2_int.get(item) != None:
    pass
```

pyright report  `case 2` as an error, but ignore `case 1`.

what if i write this codes:
```
str_2_option_int = {"1": None}    # type: dict[str, int | None]

if  str_2_int.get(item) != None:
    pass
```

how can i write type safe and correct code?

**Comment by erictraut** [comment]
Yes, this is correct. Case 1 is not a type error, but case 2 is. Mypy (another Python type checker) emits the same errors in this case.

Case 2 involves a direct call to a function with annotated parameter types, and the argument you're passing to the function violates type rules. Therefore it's an error.

Case 1 is not a direct call. It's using the `not in` operator which translates to internal behaviors of the class that are not fully specified by the type system. It is therefore not considered a type error.

**Comment by egmkang** [comment]
@erictraut 

so i cannot write type safe and correct code using `in` or `not in`?

**Comment by erictraut** [comment]
A static type checker is designed to catch certain classes of bugs, but it will always be possible to write code that does not do what you had intended. The `in` and `not int` operators will not generate runtime exceptions, so your code will run, but they might not do what you had intended if you use them incorrectly. The same is true for other operators. For example, if you use `==` when you had meant to use `!=`, your program will probably not violate any type rules, but it will not be "correct".

Consider the following example:

```python
def func(x: str):
    if x in [1, None]:
        print("Found it")
    else:
        print("Did not find it")
```

This code will always print "Did not find it". It's legal code that does not violate any type rules, but it is perhaps not what you had intended. Should a static type checker emit a warning in this case to tell you that the condition will always evaluate to `False`? Maybe, but probably not. Like I said above, if this were a very common source of bugs and the chances of a false positive were very low, it might be worth adding a warning for this condition. But my sense is that this isn't a common source of bugs, and there are legitimate uses for this pattern (especially once you move away from a trivial example like the one above). Adding a warning here would therefore generate false positives and would find very few bugs. That's not a good tradeoff.

**Comment by egmkang** [comment]
@erictraut 

python don't have a `ContainsKey` member function, and `in` or `not in` don't check Type. i add types with all my codes , but `in` expression cannot check type. it happend twice in one month, so i want to sovle this problem.

**Comment by erictraut** [comment]
Here's a potential solution that involves a custom class.

```python
T = TypeVar("T")

class Contains(Generic[T]):
    def __init__(self, iter: Iterable[T]):
        self._iter = iter

    def __call__(self, item: T) -> bool:
        return item in self._iter

if not Contains(str_2_int)(item):
    pass
```

**Comment by egmkang** [comment]
@erictraut 
```
from typing import DefaultDict, Generic, TypeVar


_KT = TypeVar("_KT")
_VT = TypeVar("_VT")


class Dict(DefaultDict[_KT, _VT], Generic[_KT, _VT]):
    def contains_key(self, key: _KT) -> bool:
        if key in self:
            return True
        return False


a = Dict[int, str]()
a[1] = "12"

print("1" in a)
print(a.contains_key("1"))
```
